server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -177,6 +179,23 @@ public DocsStats docStats() {
         }
     }
 
+    /**
+     * Returns the file sizes for the current commit
+     */
+    public List<SegmentFileSize> getLastCommittedSegmentFileSizes() {
+        SegmentInfos segmentInfos = getLastCommittedSegmentInfos();
```

Comment:
Are `getLastCommittedSegmentInfos` and `SegmentInfos` objects thread-safe?

Reply:
No, but it is used as it is.

The only place `getLastCommittedSegmentInfos()` is currently used is in `Engine.commitStats()` where it is used in the constructor of `CommitStats`.  The underlying reference is volatile in `InternalEngine` and final in `ReadOnlyEngine`.

In `SegmentInfos`, the `segment` list is modified in `applyMergeChanges`, `add`, `clear` and `remove`.

`SegmentCommitInfo` is thread safe.

Reply:
I would expect (but haven't confirmed for sure) that all modifications to `SegmentInfos#segments` happen before it is committed. If so, the value returned from `getLastCommittedSegmentInfos()` is effectively immutable :+1: